File: Java/src/main/java/QuestionBank/Solution501_600/Solution543.java
package QuestionBank.Solution501_600;

import tool.TreeNode;

/**
 * @author windows
 */
public class Solution543 {
    int max = 1;
    public int diameterOfBinaryTree(TreeNode root) {
        if(root == null){
            return 0;
        }
        if (root.left == null && root.right == null){
            return 1;
        }
        max = Math.max(max, diameterOfBinaryTree(root.left) + diameterOfBinaryTree(root.right) + 1);
        return max;
    }


    public static void main(String[] args) {
        Solution543 solution543 = new Solution543();
    }
}

File: C++/src/Solution501_600/Solution543.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution543.h"
#include <functional>
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
using namespace std;

//Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

int diameterOfBinaryTree(TreeNode* root) {
  int res = 0;
  function<int(TreeNode* root)> dfs = [&](TreeNode* root){
    if (!root){
      return 0;
    }
    int left = dfs(root->left);
    int right= dfs(root->right);
    res = max(res, left + right);
    return max(left, right) + 1;
  };
  dfs(root);
  return res;
}